src/mvba/vcbc/mod.rs

```diff
@@ -1,75 +1,249 @@
+
+                if d != &msg_d {
+                    warn!(
+                        "c-ready has unknown digest. expected {:?}, got {:?}",
+                        d, msg_d
+                    );
+                    return Err(Error::Generic("Invalid digest".to_string()));
+                }
+
+                // Upon receiving message (ID.j.s, c-ready, d, νl) from Pl for the first time:
+                if let Vacant(e) = self.wd.entry(sender) {
+                    let valid_sig = self
+                        .pub_key_set
+                        .public_key_share(sender)
+                        .verify(&sig_share, &sign_bytes);
+
+                    if !valid_sig {
+                        warn!("c-ready has has invalid signature share");
+                    }
+
+                    // if i = j and νl is a valid S1-signature share then
+                    if self.i == msg.tag.j && valid_sig {
+                        // Wd ← Wd ∪ {νl}
+                        e.insert(sig_share.clone());
+
+                        //  rd ← rd + 1
+                        self.rd += 1;
+
+                        // if rd = n+t+1/2 then
+                        if self.rd > self.threshold() {
+                            // combine the shares in Wd to an S1 -threshold signature µ
+                            let sig = self.pub_key_set.combine_signatures(self.wd.iter())?;
+
+                            let final_msg = Message {
+                                tag: self.tag.clone(),
+                                action: Action::Final(d.clone(), sig),
+                            };
+
+                            // send (ID.j.s, c-final, d, µ) to all parties
+                            self.broadcast(&final_msg)?;
+                        }
+                    }
+                }
+            }
+            Action::Final(msg_d, sig) => {
+                // Upon receiving message (ID.j.s, c-final, d, µ):
+                let d = match &self.d {
+                    Some(d) => d,
+                    None => {
+                        warn!("received c-final before receiving s-send, logging message");
+                        try_insert(&mut self.final_messages, sender, msg)?;
+                        return Ok(());
+                    }
+                };
+
+                let sign_bytes = self.c_ready_bytes_to_sign(d)?;
+                let valid_sig = self.pub_key_set.public_key().verify(&sig, sign_bytes);
+
+                if !valid_sig {
+                    warn!("c-ready has has invalid signature share");
+                }
+
+                // if H(m̄) = d and µ̄ = ⊥ and µ is a valid S1 -signature then
+                if d == &msg_d && self.u_bar.is_none() && valid_sig {
+                    // µ̄ ← µ
+                    self.u_bar = Some(sig);
+                }
+            }
+        }
+
+        for (sender, final_msg) in std::mem::take(&mut self.final_messages) {
```

Comment:
I prefer keep all messages in buffer and then process them. Other wise we need put some comments here to explain why we only buffered the final messages.